Polymorphic varargs methods fail

When a variable argument formal parameter is annotated with the polymorphic type, it should be possible to pass arguments annotated with any type in the type hierarchy in any order. It looks like the checker framework incorrectly resolves the polymorphic type to the type of the first argument in the variable argument list, instead of taking their least upper bound.
Consider the following example to be checked with the nullness checker:

```
import org.checkerframework.checker.nullness.qual.*;

public class Issue24PolyNull {
    static void varArgsMethod(@PolyNull Object... args) {
    }
    static void callToVarArgsObject(@PolyNull Object a, @NonNull Object b) {
        varArgsMethod(b, a);
        varArgsMethod(a, b);
    }
}

```

We expect this code to type check without errors. But, executing the command `javac -processor nullness Issue24PolyNull.java` results in the following error:

```
Issue24PolyNull.java:9: error: [argument.type.incompatible] incompatible types in argument.
        varArgsMethod(b, a);
                         ^
  found   : @Initialized @PolyNull Object
  required: @Initialized @NonNull Object
```
Note that the call `varArgsMethod(a, b)` type checks correctly. It appears that for variable arguments, the type of the first argument is used to resolve the polymorphic type. For `varArgsMethod(b, a);`, `@PolyNull` resolves to `@NonNull` causing the error. For `varArgsMethod(a, b);`, `@PolyNull` resolves to `@PolyNull` correctly. So the solution would be for the checker framework to take the least upper bound of all arguments in the variable argument list while resolving such polymorphic annotations.

Here is an expanded test case

```
import org.checkerframework.checker.nullness.qual.*;

public class Issue2171 {
    static void varArgsMethod(@PolyNull Object... args) {}

    static void callToVarArgsObject(@PolyNull Object pn, @NonNull Object nn, @Nullable Object nble) {
        varArgsMethod(nble, nble);
        varArgsMethod(nble, nn);
        varArgsMethod(nble, pn);
        varArgsMethod(nn, nble);
        varArgsMethod(nn, nn);
        varArgsMethod(nn, pn);
        varArgsMethod(pn, nble);
        varArgsMethod(pn, nn);
        varArgsMethod(pn, pn);
    }
}
```

Here is the output:

```
$ch/bin/javac -g Issue2171.java -processor nullness
Issue2171.java:10: error: [argument.type.incompatible] incompatible types in argument.
        varArgsMethod(nn, nble);
                          ^
  found   : @Initialized @Nullable Object
  required: @Initialized @NonNull Object
Issue2171.java:12: error: [argument.type.incompatible] incompatible types in argument.
        varArgsMethod(nn, pn);
                          ^
  found   : @Initialized @PolyNull Object
  required: @Initialized @NonNull Object
Issue2171.java:13: error: [argument.type.incompatible] incompatible types in argument.
        varArgsMethod(pn, nble);
                          ^
  found   : @Initialized @Nullable Object
  required: @Initialized @PolyNull Object
3 errors
```

--- a/checker/tests/nullness/Issue2171.java
+++ b/checker/tests/nullness/Issue2171.java
@@ -1,3 +1,4 @@
+import java.util.List;
 import org.checkerframework.checker.nullness.qual.*;
 
 public class Issue2171 {
@@ -14,5 +15,20 @@ public class Issue2171 {
         varArgsMethod(pn, nble);
         varArgsMethod(pn, nn);
         varArgsMethod(pn, pn);
+    }
+
+    static void genVarArgsMethod(List<? extends @PolyNull Object>... args) {}
+
+    static void genCallToVarArgsObject(
+            List<@PolyNull Object> pn, List<@NonNull Object> nn, List<@Nullable Object> nble) {
+        genVarArgsMethod(nble, nble);
+        genVarArgsMethod(nble, nn);
+        genVarArgsMethod(nble, pn);
+        genVarArgsMethod(nn, nble);
+        genVarArgsMethod(nn, nn);
+        genVarArgsMethod(nn, pn);
+        genVarArgsMethod(pn, nble);
+        genVarArgsMethod(pn, nn);
+        genVarArgsMethod(pn, pn);
     }
 }

--- a/checker/tests/nullness/Issue2171.java
+++ b/checker/tests/nullness/Issue2171.java
@@ -17,8 +17,10 @@ public class Issue2171 {
         varArgsMethod(pn, pn);
     }
 
+    @SuppressWarnings("unchecked")
     static void genVarArgsMethod(List<? extends @PolyNull Object>... args) {}
 
+    @SuppressWarnings("unchecked")
     static void genCallToVarArgsObject(
             List<@PolyNull Object> pn, List<@NonNull Object> nn, List<@Nullable Object> nble) {
         genVarArgsMethod(nble, nble);

--- a/checker/tests/nullness/Issue2171.java
+++ b/checker/tests/nullness/Issue2171.java
@@ -1,0 +1,18 @@
+import org.checkerframework.checker.nullness.qual.*;
+
+public class Issue2171 {
+    static void varArgsMethod(@PolyNull Object... args) {}
+
+    static void callToVarArgsObject(
+            @PolyNull Object pn, @NonNull Object nn, @Nullable Object nble) {
+        varArgsMethod(nble, nble);
+        varArgsMethod(nble, nn);
+        varArgsMethod(nble, pn);
+        varArgsMethod(nn, nble);
+        varArgsMethod(nn, nn);
+        varArgsMethod(nn, pn);
+        varArgsMethod(pn, nble);
+        varArgsMethod(pn, nn);
+        varArgsMethod(pn, pn);
+    }
+}

--- a/framework/src/main/java/org/checkerframework/framework/util/AnnotatedTypes.java
+++ b/framework/src/main/java/org/checkerframework/framework/util/AnnotatedTypes.java
@@ -629,7 +629,7 @@ public class AnnotatedTypes {
 
         parameters = new ArrayList<>(parameters.subList(0, parameters.size() - 1));
         for (int i = args.size() - parameters.size(); i > 0; --i) {
-            parameters.add(varargs.getComponentType());
+            parameters.add(varargs.getComponentType().shallowCopy());
         }
 
         return parameters;

--- a/framework/src/main/java/org/checkerframework/framework/util/AnnotatedTypes.java
+++ b/framework/src/main/java/org/checkerframework/framework/util/AnnotatedTypes.java
@@ -629,7 +629,7 @@ public class AnnotatedTypes {
 
         parameters = new ArrayList<>(parameters.subList(0, parameters.size() - 1));
         for (int i = args.size() - parameters.size(); i > 0; --i) {
-            parameters.add(varargs.getComponentType().shallowCopy());
+            parameters.add(varargs.getComponentType().deepCopy());
         }
 
         return parameters;

